roofit/roofitcore/src/RooAbsPdf.cxx

```diff
@@ -1496,7 +1510,49 @@ RooFitResult* RooAbsPdf::fitTo(RooAbsData& data, const RooLinkedList& cmdList)
 	// Evaluate errors with Hesse
 	m.hesse() ;
       }
-      
+
+      //asymptotically correct approach
+      if (doAsymptotic==1 && m.getNPar()>0) {
+	// Calculated corrected errors for weighted likelihood fits
+	std::unique_ptr<RooFitResult> rw(m.save());
+	//weighted inverse Hessian matrix
+	const TMatrixDSym& matV = rw->covarianceMatrix();
+	coutI(Fitting) << "RooAbsPdf::fitTo(" << GetName() << ") Calculating covariance matrix according to the asymptotically correct approach. If you find this method useful please consider citing https://arxiv.org/abs/1911.01303." << endl;
+	//initialise matrix containing first derivatives
+	TMatrixDSym num(rw->floatParsFinal().getSize());
+	for (int k=0; k<rw->floatParsFinal().getSize(); k++)
+	   for (int l=0; l<rw->floatParsFinal().getSize(); l++)
+	      num(k,l) = 0.0;
+	//determine matrix containing first derivatives
+	RooArgSet* obs = getObservables(data);      
+	for (int j=0; j<data.numEntries(); j++) {
+	   *obs = *data.get(j);//sets obs to current data point, this is where the pdf will be evaluated
+	   std::unique_ptr<RooArgSet> floatingparams( (RooArgSet*)getParameters(data)->selectByAttrib("Constant", false) );
+	   const RooArgList& floated = rw->floatParsFinal();
+	   //determine first derivatives
+	   vector<Double_t> diffs(floated.getSize(), 0.0);
+	   for (int k=0; k<floated.getSize(); k++) {
+	      RooRealVar* paramresult = (RooRealVar*)floated.at(k);
+	      RooRealVar* paraminternal = (RooRealVar*)floatingparams->find(paramresult->getTitle());
+	      //determine first derivative to parameter k at best estimate point
+	      std::unique_ptr<RooDerivative> deriv( derivative(*paraminternal, *obs, 1) );
```

Comment:
Only when going through the code a second time, we realised that constructing and destroying the RooDerivative *for every point and parameter* is probably inefficient. It should be possible to create one derivative object for each parameter before the loop, save them in a `std::vector` or so, and to evaluate each of them at any given data point.